android/runtime/common/src/java/org/appcelerator/kroll/KrollRuntime.java

```diff
@@ -541,6 +558,59 @@ public static void dispatchException(final String title, final String message, f
 		}
 	}
 
+	/**
+	 * Adds a listener to be invoked just before a Titanium JavaScript runtime instance has been terminated.
+	 * <p>
+	 * You cannot add the same listener instance twice. Duplicate listener instances will be ignored.
+	 * @param listener The listener to be added. Null references will be ignored.
+	 */
+	public static void addOnDisposingListener(KrollRuntime.OnDisposingListener listener)
+	{
+		if ((listener != null) && (disposingListeners.contains(listener) == false)) {
+			disposingListeners.add(listener);
+		}
+	}
+
+	/**
+	 * Removes the given listener by reference that was added via the addOnDisposingListener() method.
+	 * @param listener The listener to be removed by reference. Can be null.
+	 */
+	public static void removeOnDisposingListener(KrollRuntime.OnDisposingListener listener)
+	{
+		if (listener != null) {
+			disposingListeners.remove(listener);
+		}
+	}
+
+	/**
+	 * This method is intended to be called just before the given runtime's doDispose() method has been called.
+	 * Invokes all "OnDisposingListener" objects, allowing them to perform final cleanup operations.
+	 * @param runtime The runtime instance that is about to be disposed of. Can be null.
+	 */
+	private static void onDisposing(KrollRuntime runtime)
+	{
+		// Validate.
+		if (runtime == null) {
+			return;
+		}
+
+		// Create a shallow copy of all listeners to be iterated down below.
+		// We do this because an invoked listener can add/remove listeners in main collection.
+		ArrayList<OnDisposingListener> clonedListeners = null;
```

Comment:
Why not assign `cloudListeners` right away?
```Java
ArrayList<OnDisposingListener> clonedListeners = (ArrayList<OnDisposingListener>) KrollRuntime.disposingListeners.clone();
```
I know the line is pretty long, but our linter will format it